### ISSUE TITLE
[Bug] High GPU memory consumption of ```ProductKernel``` compared to its equivalent variant

### ISSUE THREAD

**Issue by patel-zeel** [cornellius-gp/gpytorch#1806]
# 🐛 Bug

I want to apply different kernels on different dimensions for a specific problem, but, I am running out of GPU memory when using GPyTorch. Narrowing down the problem to its simplest version: 
* Using products of multiple ```RBFKernel``` on each dimension takes more GPU memory compared to a single ```RBFKernel``` with ```active_dims``` argument.
* Memory consumption difference increases with more dimensions.

## To reproduce

```python
import torch
import gpytorch


class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood, kernel):
        super(ExactGPModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = kernel

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)


likelihood = gpytorch.likelihoods.GaussianLikelihood().cuda()

n_observations = 10000
dims = 3

train_x = torch.rand(n_observations, dims).cuda()
train_y = (
    torch.sin(train_x[:, 0] * 5)
    + torch.sin(train_x[:, 1] * 10)
    + torch.sin(train_x[:, 2] * 15)
).cuda()

### Version 1
# K = None
# for i in range(dims):
#     if K is None:
#         K = gpytorch.kernels.RBFKernel(ard_num_dims=1, active_dims=[i])
#     else:
#         K = K * gpytorch.kernels.RBFKernel(ard_num_dims=1, active_dims=[i])

### Version 2
K = gpytorch.kernels.RBFKernel(ard_num_dims=dims, active_dims=[0, 1, 2])

model = ExactGPModel(
    train_x, train_y, likelihood, kernel=gpytorch.kernels.ScaleKernel(K)
).cuda()
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

model.train()
likelihood.train()

optimizer = torch.optim.Adam(model.parameters(), lr=0.1)

training_iter = 100
for i in range(training_iter):
    optimizer.zero_grad()
    output = model(train_x)
    loss = -mll(output, train_y)
    loss.backward()
    print(
        i + 1,
        training_iter,
        loss.item(),
        model.covar_module.base_kernel.lengthscale,
        model.likelihood.noise.item(),
    )
    optimizer.step()
```

## Expected Behavior

For ```### Version 1``` and ```### Version 2``` kernels, GPU should consume the same amount of GPU memory but currently, it consumes 5069 MB for ```### Version 1``` and 3161 MB for ```### Version 2```.

## System information

-  GPyTorch 1.5.1
-  PyTorch 1.9.0+cu102
-  Ubuntu 18.04

**Comment by gpleiss** [comment]
This is expected behavior. Think about the computation graphs associated with these two versions.
- Version 1: the `ProductKernel` module creates 3 kernel matrices (all of which require gradients), and then multiplies them together. Since the 3 kernel matrices require gradients, they will remain instantiated until you call .backward.
- Version 2: you are only instantiating one kernel matrix.

So it is totally expected that version 2 is more memory efficient.

**Comment by patel-zeel** [comment]
Thank you for clarifying the differences @gpleiss. I was then wondering how to implement something like ```CategoricalKernel()``` without eating up extra GPU memory? [BoTorch version](https://botorch.org/api/_modules/botorch/models/kernels/categorical.html) of this kernel would take the same amount of memory as ```ProductKernel()```? I can see that the following line will require at least ```n * m * d``` memory to store the distance matrix for ```n * d``` sized x1 and ```m * d``` sized x2.

```python
delta = x1.unsqueeze(-2) != x2.unsqueeze(-3)
```

**Comment by patel-zeel** [comment]
Seems exciting! Shall I open a PR if I am able to construct a ```CategoricalKernel``` by internally using ```IndexKernel``` ?

**Comment by gpleiss** [comment]
Sure!
